### PR TITLE
Refactoring the `ChessBoardState`

### PR DESCRIPTION
**Pull request by KurohanaJuri**
# Problematic
At this time, the `ChessBoardState` is too specific to the 2D board (aka contains methods used to move pieces). This means that cannot be used by the 3D board

# Description
This PR is the 1st part of #239 and his goal is to extract common information. The new AR using the new state will be done in a separate PR.

### CODE REVIEW

**Comment by alexandrepiveteau** [comment]
I don't think using generic `Rank` and `Color` types will is the right approach, since **it won't let us reuse the implementation of the `ChessBoardState` across both the classic and AR modes if the ranks and color union types are not shared**. Instead, I think we should proceed as follows :

1. Extract the `contentDescription`, `whiteIcon` and `blackIcon` properties from `Rank` and move them to a dedicated function with a `when` clause, called from the `ChessBoard` composable.
2. Extract the `contentDescription` property from `Color` and move it to a dedicated function with a `when` clause, called from the `ChessBoard` composable.
3. Create a `MoveableChessBoardState` interface which extends `ChessBoardState` and defines the "moveable" aspects of the board. Update the `ChessBoard` composable signature to use a `MoveableChessBoardState`.

The final interfaces would look similar to this :

```kotlin
@Stable
interface ChessBoardState<out Piece : ChessBoardState.Piece> {

  enum class Rank {
    King,
    Queen,
    Rook,
    Bishop,
    Knight,
    Pawn,
  }

  enum class Color {
    Black,
    White,
  }

  data class Position(val x: Int, val y: Int)

  @Stable
  interface Piece {
    val rank: Rank
    val color: Color
  }

  val pieces: Map<Position, Piece>
  val checkPosition: Position?
}
```

```kotlin
@Stable
interface MoveableChessBoardState<Piece : ChessBoardState.Piece> {
  val selectedPosition: ChessBoardState.Position?
  val availableMoves: Set<ChessBoardState.Position>
  fun onDropPiece(piece: Piece, endPosition: ChessBoardState.Position)
  fun onPositionClick(position: ChessBoardState.Position)
}
```

The `ChessBoard` composable could then be renamed to `ClassicChessBoard`, and the AR composable could be renamed to `ArChessBoard` too for instance.

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/GameScreenState.kt [hunk 1]
```diff
@@ -9,7 +9,7 @@ import androidx.compose.runtime.Stable
  * @param Piece the type of the pieces of the underlying [ChessBoardState].
  */
 @Stable
-interface GameScreenState<Piece : ChessBoardState.Piece> : ChessBoardState<Piece> {
+interface GameScreenState<Pieces : ChessBoardState.Piece> : MovableChessBoardState<Pieces> {
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/GameScreenState.kt:12]:
```suggestion
interface GameScreenState<Piece : ChessBoardState.Piece> : MovableChessBoardState<Piece> {
```

Comment (KurohanaJuri) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/GameScreenState.kt]: Fixed in commit https://github.com/epfl-SDP/android/pull/240/commits/e937b3eef4a540d698d097c248e35b2ba29c6dd3

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/MovableChessBoardState.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package ch.epfl.sdp.mobile.ui.game
+
+import androidx.compose.runtime.Stable
+
+/**
+ * An interface that extend the interface [ChessBoardState] and offers some ways for the
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/MovableChessBoardState.kt:6]:
```suggestion
 * An interface that extends the interface [ChessBoardState] and offers some ways for the
```

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/ClassicChessBoard.kt [hunk 3]
```diff
@@ -216,20 +219,66 @@ private fun Piece(
...
+  }
+}
+
+/** Returns the [String] content description associated to the value of this [Piece]. */
+private val Piece.contentDescription: String
+  @Composable
+  get() {
+    val strings = LocalLocalizedStrings.current
+    val color = color.contentDescription(strings)
+
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/ClassicChessBoard.kt:280]:
```suggestion
```

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/ClassicChessBoard.kt [hunk 4]
```diff
@@ -216,20 +219,66 @@ private fun Piece(
...
+    val rank = rank.contentDescription(strings)
+
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/ClassicChessBoard.kt:282]:
```suggestion
```

**Review by alexandrepiveteau — COMMENTED** [review]
Overall LGTM. There are a few details (esp. when it comes to renaming some classes) but they should be straightforward to address. Thanks !

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/MovableChessBoardState.kt (hunk 2 above)
Comment (KurohanaJuri) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/MovableChessBoardState.kt]: Fix in commit https://github.com/epfl-SDP/android/pull/240/commits/23fd568305d5eeeb1ea1c92c3947b89dcb54407a

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/ChessBoardState.kt [hunk 5]
```diff
@@ -1,91 +1,44 @@
...
-import ch.epfl.sdp.mobile.ui.*
-import ch.epfl.sdp.mobile.ui.game.ChessBoardState.Piece
-import ch.epfl.sdp.mobile.ui.i18n.LocalizedStrings
 
 /**
- * An interface which represents the state of a [ChessBoard] composable. It display the current
- * state of the game, and offers some ways for the [ChessBoard] to interact with it and indicate the
- * user gestures.
+ * An interface which represents the state of a [ClassicChessBoard] composable. It display the
+ * current state of the game
```
[the remark this answers is not in the document]
Comment (KurohanaJuri) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/ChessBoardState.kt]: Fix in commit https://github.com/epfl-SDP/android/pull/240/commits/23fd568305d5eeeb1ea1c92c3947b89dcb54407a

File: mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulGameScreen.kt [hunk 6]
```diff
@@ -78,7 +76,7 @@ fun StatefulGameScreen(
  * @param match the match to display.
  * @param scope a [CoroutineScope] keeping track of the state lifecycle.
  */
-class SnapshotChessBoardState(
+class SnapshotClassicChessBoardState(
```
[the remark this answers is not in the document]
Comment (KurohanaJuri) [reply mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulGameScreen.kt]: Fix in commit https://github.com/epfl-SDP/android/pull/240/commits/a1b13dd233acbf3f82e412eaf9a9e283b75c4bbd

File: mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/state/ClassicChessBoardStateTest.kt [hunk 7]
```diff
@@ -40,7 +40,7 @@ class ChessBoardStateTest {
     val match = facade.createMatch(user, user)
 
     val actions = StatefulGameScreenActions(onBack = {}, onShowAr = {})
-    val state = SnapshotChessBoardState(actions, user, match, scope)
+    val state = SnapshotClassicChessBoardState(actions, user, match, scope)
```
[the remark this answers is not in the document]
Comment (KurohanaJuri) [reply mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/state/ClassicChessBoardStateTest.kt]: https://github.com/epfl-SDP/android/pull/240/commits/a1b13dd233acbf3f82e412eaf9a9e283b75c4bbd

**Review by alexandrepiveteau — DISMISSED** [review]
Thanks !

**Review by alexandrepiveteau — DISMISSED** [review]

**Review by matt989253 — APPROVED** [review]
Nice, looks very clean!

**Review by alexandrepiveteau — APPROVED** [review]

**Pull request merged**